compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/SamplingTaskSizingPass.java

```diff
@@ -0,0 +1,468 @@
+          TaskSizeSplitterVertex prevSplitter = (TaskSizeSplitterVertex) incomingEdge.getSrc();
+          IREdge internalEdge = prevSplitter.getEdgeWithInternalVertex(incomingEdge);
+          IREdge newIrEdge = Util.cloneEdge(incomingEdge, incomingEdge.getSrc(), toInsert);
+          prevSplitter.mapEdgeWithLoop(newIrEdge, internalEdge);
+          fromOutsideToSplitter.add(newIrEdge);
+        } else {
+          IREdge cloneOfIncomingEdge = Util.cloneEdge(incomingEdge, incomingEdge.getSrc(), toInsert);
+          fromOutsideToSplitter.add(cloneOfIncomingEdge);
+        }
+      }
+    }
+    return fromOutsideToSplitter;
+  }
+
+  /**
+   * Set edges which go out from splitter vertex to other stages.
+   * @param dag                             dag to modify
+   * @param toInsert                        splitter vertex to check
+   * @param verticesWithStageOutgoingEdges  vertices with stage outgoing edges
+   * @return
+   */
+  private Set<IREdge> setEdgesFromSplitterToOutside(final IRDAG dag,
+                                                    final TaskSizeSplitterVertex toInsert,
+                                                    final Set<IRVertex> verticesWithStageOutgoingEdges) {
+    HashSet<IREdge> fromSplitterToOutside = new HashSet<>();
+    for (IRVertex vertex : verticesWithStageOutgoingEdges) {
+      for (IREdge outgoingEdge : dag.getOutgoingEdgesOf(vertex)) {
+        if (outgoingEdge.getDst() instanceof TaskSizeSplitterVertex) {
+          TaskSizeSplitterVertex nextSplitter = (TaskSizeSplitterVertex) outgoingEdge.getDst();
+          IREdge internalEdge = nextSplitter.getEdgeWithInternalVertex(outgoingEdge);
+          IREdge newIrEdge = Util.cloneEdge(outgoingEdge, toInsert, outgoingEdge.getDst());
+          nextSplitter.mapEdgeWithLoop(newIrEdge, internalEdge);
+          fromSplitterToOutside.add(newIrEdge);
+        } else if (!toInsert.getOriginalVertices().contains(outgoingEdge.getDst())) {
+          IREdge cloneOfOutgoingEdge = Util.cloneEdge(outgoingEdge, toInsert, outgoingEdge.getDst());
+          fromSplitterToOutside.add(cloneOfOutgoingEdge);
+        }
+      }
+    }
+    return fromSplitterToOutside;
+  }
+
+  /**
+   * Make splitter vertex and insert it in the dag.
+   * @param dag                              dag to insert splitter vertex
+   * @param stageVertices                    stage vertices which will be grouped to be inserted into splitter vertex
+   * @param stageStartingVertices            subset of stage vertices which have incoming edge from other stages
+   * @param verticesWithStageOutgoingEdges   subset of stage vertices which have outgoing edge to other stages
+   * @param stageEndingVertices              subset of staae vertices which does not have outgoing edge to other
+   *                                         vertices in this stage
+   * @param partitionerProperty              partitioner property
+   */
+  private void makeAndInsertSplitterVertex(final IRDAG dag,
```

Comment:
`insertSplitterVertex` seems enough for the naming

Reply:
You have a point. Thank you.